File: src/observable/observable_base.py
import abc
import uuid
from typing import TYPE_CHECKING, Any, Mapping
from weakref import WeakKeyDictionary

from ..codegen import HasCodegenBuiltin
from ..comm import ReceiverFn_t
from .messages import OBSERVABLE_CHANGE_MSG, BaseChangeMessage

if TYPE_CHECKING:
    from .observable_network import ObservableNetwork


__all__ = ["ObservableBase"]


class ObservableBase[Schema_T: Mapping[str, Any], Message_T: BaseChangeMessage](
    HasCodegenBuiltin, abc.ABC
):
    __network: "ObservableNetwork | None"
    __id: str
    __receivers: WeakKeyDictionary[ReceiverFn_t[Any], ReceiverFn_t[Any]]

    def __init__(self) -> None:
        super().__init__()

        self.__network = None
        self.__id = uuid.uuid4().hex
        self.__receivers = WeakKeyDictionary()

    @classmethod
    def __codegen_builtin__(cls) -> str:
        return cls.__name__

    @property
    def id(self) -> str:
        return self.__id

    @property
    def network(self) -> "ObservableNetwork | None":
        return self.__network

    def watch(self, __fn: ReceiverFn_t[Message_T]) -> ReceiverFn_t[Message_T]:
        if __fn in self.__receivers:
            return __fn

        def __recv_watch(__change: Message_T) -> None:
            if __change["data_id"] != self.id:
                return
            __fn(__change)

        self.__receivers[__fn] = __recv_watch

        if self.__network is not None:
            self.__network.comm.recv(OBSERVABLE_CHANGE_MSG, __recv_watch)

        return __fn

    def unwatch(self, __fn: ReceiverFn_t[Message_T]) -> ReceiverFn_t[Message_T]:
        if __fn in self.__receivers:
            if self.__network is not None:
                self.__network.comm.unrecv(
                    OBSERVABLE_CHANGE_MSG, self.__receivers[__fn]
                )
            del self.__receivers[__fn]
        return __fn

    def _register(self, __network: "ObservableNetwork") -> None:
        self.__network = __network
        for receiver in self.__receivers.values():
            self.__network.comm.recv(OBSERVABLE_CHANGE_MSG, receiver)

    @abc.abstractmethod
    def load_schema(self, schema: Schema_T) -> None:
        ...

    @abc.abstractmethod
    def __json__(self) -> Schema_T:
        ...

    @abc.abstractmethod
    def _handle_client_change(self, change: Message_T) -> None:
        ...

File: src/observable/observable_base.ts
import { ReceiverFn_t } from "../comm/__init__.js";
import { weakref } from "../utils/__init__.js";
import { get_default } from "../utils/dicttools.js";
import { uuid4 } from "../utils/uuid.js";
import { BaseChangeMessage, OBSERVABLE_CHANGE_MSG } from "./messages.js";
import { ObservableNetwork } from "./observable_network.js";
import { BaseObservableSchema } from "./schema.js";

/**
 * The keywords supported by {@link ObservableBase}
 */
export type ObservableKwds = {
    id?: string;
    remote?: boolean;
    network?: ObservableNetwork;
};

export abstract class ObservableBase<Schema_T extends BaseObservableSchema, Message_T extends BaseChangeMessage> {
    #id: string;
    #remote: boolean;
    #network?: ObservableNetwork;
    #receivers: weakref.IterableWeakMap<ReceiverFn_t, ReceiverFn_t>;
    #is_updating: boolean;

    protected constructor(kwds: ObservableKwds = {}) {
        this.#id = get_default(kwds, "id", uuid4());
        this.#remote = get_default(kwds, "remote", false);
        this.#network = get_default(kwds, "network", undefined);
        this.#receivers = new weakref.IterableWeakMap();
        this.#is_updating = false;
    }

    /**
     * Gets the {@link Observable}'s id
     *
     * @readonly
     */
    get id(): string {
        return this.#id;
    }

    /**
     * Gets the {@link Observable}'s remote property
     *
     * @readonly
     */
    get remote(): boolean {
        return this.#remote;
    }


    /**
     * Gets the {@link Observable}'s {@link ObservableNetwork}
     *
     * @readonly
     */
    get network(): ObservableNetwork | void {
        return this.#network;
    }

    /**
     * Gets the {@link Observable}'s {@link ObservableNetwork}
     *
     * @readonly
     */
    get receivers(): weakref.IterableWeakMap<ReceiverFn_t, ReceiverFn_t> {
        return this.#receivers;
    }

    /**
     * Watches the observable and calls the callback with its value
     * as the argument when changed.
     *
     * @param {TransformFn<T, any>} fn the callback
     * @see {@link Observable.watch}
     */
    watch(fn: (change: Message_T) => any): void {
        const __recv_change = (change: Message_T) => {
            if (change["data_id"] != this.id) { return; }
            fn(change);
        };

        this.receivers.set(fn, __recv_change);
        this.network?.comm.recv(OBSERVABLE_CHANGE_MSG, __recv_change);
    }

    /**
     * Internal method: used by {@link ObservableNetwork} as part of registration
     *
     * @private
     * @param {ObservableNetwork} network
     */
    __register(network: ObservableNetwork): void {
        this.#network = network;
        for (let fn of this.#receivers.values()) {
            this.#network.comm.recv(OBSERVABLE_CHANGE_MSG, fn);
        }
    }


    /**
     * Internal method: used by {@link ObservableNetwork} as part of communication
     *
     * @private
     * @param change
     */
    __recv_change(change: Message_T): void {
        this._handle_change_message(change);
    }

    protected _emit_change_message(change: Message_T, local: boolean = false): void {
        if (this.#is_updating) { return; }

        this.#is_updating = true;

        if (this.remote && !local) {
            this.network?.send_change(change);
        }
        else {
            this.network?.notify_change(change);
        }

        this.#is_updating = false;
    }

    protected abstract _handle_change_message(msg: Message_T): void;
}

